### PR TITLE
Feat/new pan view with reanimated 2

### PR DESCRIPTION
**Pull request by M-i-k-e-l**
## Description
New unified PanView - initial commit to incubator.
TODOs:
1. Hidden location (where will the view animate to when dismissed) is not calculated on layout --> very long time until `onDismiss` is called.
2. Need to add enter animation.

## Changelog
New unified PanView - initial commit to incubator.

### CODE REVIEW

**Review by ethanshar — COMMENTED** [review]
Looks and works great! 
Wrote some comments..

File: demo/src/screens/incubatorScreens/PanViewScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+  TouchableOpacity
+} from 'react-native-ui-lib';
+const {PanView} = Incubator;
+
+interface Item {
+  value: string;
+  label: string;
+}
+
+const chevronDown = require('../../assets/icons/chevronDown.png');
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply demo/src/screens/incubatorScreens/PanViewScreen.tsx:23]: Fixed

File: src/incubator/index.ts [hunk 2]
```diff
@@ -3,3 +3,4 @@ export {default as TabController} from './TabController';
 export {default as TextField, TextFieldProps, FieldContextType} from './TextField';
 export {default as TouchableOpacity, TouchableOpacityProps} from './TouchableOpacity';
 export {default as WheelPicker, WheelPickerProps} from './WheelPicker';
+export {default as PanView, PanViewProps, PanningDirections, PanDismissThreshold} from './panView';
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/index.ts:6]: I agree (the `PanningDirections` already exists in `PanningProvider`, but it should be clearer in future usage).

File: src/incubator/panView/index.tsx [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  directionLock?: boolean;
+  /**
+   * Allows locking the translation when dragging or dropping.
+   * i.e. cannot drag back when a certain direction is not allowed.
+   * Only when dismissible={false}
+   */
+  translationLock?: TranslationLock;
+  /**
+   * Object to adjust the dismiss threshold limits.
+   */
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:54]: Done (you were missing `).` so I didn't commit the suggestion)

File: src/incubator/panView/index.tsx [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+PanView.displayName = 'PanView';
+PanView.directions = PanningDirections;
+PanView.translationLock = TranslationLock;
+PanView.defaultProps = {
+  directions: [PanningDirections.UP, PanningDirections.DOWN, PanningDirections.LEFT, PanningDirections.RIGHT],
+  translationLock: TranslationLock.NONE,
+  threshold: DEFAULT_THRESHOLD
+};
+
+export default asBaseComponent<PanViewProps, StaticMembers>(PanView);
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:188]: TBH, I took it from [your code](https://github.com/wix/react-native-ui-lib/blob/master/src/incubator/TextField/index.tsx#L220) ), and I think it's more explicit; another positive aspect is: we're not adding `displayName` and `defaultProps` to the statics, WDYT?

File: src/incubator/panView/index.tsx [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+          translationX.value = withSpring(0, RETURN_ANIMATION_SPRING_CONFIG);
+          translationY.value = withSpring(0, RETURN_ANIMATION_SPRING_CONFIG);
+        }
+      }
+    }
+  },
+  [directions, dismissible, setTranslation]);
+
+  return (
+    // <View ref={containerRef} style={containerStyle} onLayout={onLayout}>
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:163]: That will probably be needed in the future, for improving the animations.

File: src/incubator/panView/index.tsx [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    <View style={containerStyle}>
+      <PanGestureHandler onGestureEvent={isEmpty(directions) ? undefined : onGestureEvent}>
+        <Animated.View
+          // !visible.current && styles.hidden is done to fix a bug is iOS
+          //   style={[style, animatedStyle, !visible.current && styles.hidden]}
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:168]: This was a fix for a bug on iOS, which currently doe not happen, I'll add TODOs for this comments so they won't be left there (hopefully ;) )

File: src/incubator/panView/index.tsx [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+interface StaticMembers {
+  directions: typeof PanningDirections;
+  translationLock: typeof TranslationLock;
+}
+
+interface Props extends PanViewProps {
+  children?: React.ReactNode | React.ReactNode[];
+}
+
+const RETURN_ANIMATION_SPRING_CONFIG = {velocity: 300, damping: 15, stiffness: 300, mass: 0.8};
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:71]: Done

File: src/incubator/panView/index.tsx [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+import {StyleSheet, StyleProp, ViewStyle} from 'react-native';
+import {PanGestureHandler, PanGestureHandlerEventPayload} from 'react-native-gesture-handler';
+import {asBaseComponent} from '../../commons/new';
+import Animated, {
+  useSharedValue,
+  useAnimatedStyle,
+  withSpring,
+  useAnimatedGestureHandler,
+  runOnJS
+} from 'react-native-reanimated';
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:12]: Oops

File: src/incubator/panView/index.tsx [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+interface Props extends PanViewProps {
+  children?: React.ReactNode | React.ReactNode[];
+}
+
+const RETURN_ANIMATION_SPRING_CONFIG = {velocity: 300, damping: 15, stiffness: 300, mass: 0.8};
+
+const PanView = (props: Props) => {
+  const {
+    directions,
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:75]: That's not me, typescript should understand `defaultProps`...
I prefer to either use `defaultProps` or give a default value, WDYT?

**Comment by M-i-k-e-l** [comment]
I did not respond to all the comments, but I'd appreciate it if you can reply to the ones I did 🙏

File: src/incubator/panView/panningUtil.ts [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+/**
+ * Will return undefined if should not dismiss
+ */
+export function getDismissVelocity(event: PanGestureHandlerEventPayload,
+  directions: PanningDirections[],
+  options: TranslationOptions,
+  threshold?: PanDismissThreshold) {
+  'worklet';
+  // Sadly using {...DEFAULT_THRESHOLD, ...threshold} is done
+  // on the JS thread for some reason, causing an exception
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/panningUtil.ts]: Fixed, used `{}` as the target, hope that's ok.

File: src/components/modal/index.tsx [hunk 11]
```diff
@@ -1,5 +1,6 @@
 import _ from 'lodash';
 import React, {Component} from 'react';
+import {GestureHandlerRootView} from 'react-native-gesture-handler';
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/modal/index.tsx]: Moved to screen

File: src/incubator/panView/index.tsx [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+  /**
+   * Should the direction of dragging be locked once a drag has started.
+   */
+  directionLock?: boolean;
+  /**
+   * Allows locking the translation when dragging or dropping.
+   * i.e. cannot drag back when a certain direction is not allowed.
+   * Only when dismissible={false}
+   */
+  translationLock?: TranslationLock;
```
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/incubator/panView/index.tsx:51]: @M-i-k-e-l ?

File: src/incubator/panView/index.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/incubator/panView/index.tsx:189]: I prefer using typeof cause then you don't need to think of adding stuff to Static typings every time, it's less error prone. 
Btw, the reference you linked, Im currently having typings issues there I need to solve 😬

File: src/incubator/panView/index.tsx [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+  directions: [PanningDirections.UP, PanningDirections.DOWN, PanningDirections.LEFT, PanningDirections.RIGHT],
+  translationLock: TranslationLock.NONE,
+  threshold: DEFAULT_THRESHOLD
+};
+
+export default asBaseComponent<PanViewProps, StaticMembers>(PanView);
+
+const styles = StyleSheet.create({
+  wrapChild: {
+    alignSelf: 'baseline'
```
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/incubator/panView/index.tsx:193]: @M-i-k-e-l ?

File: src/incubator/panView/index.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/incubator/panView/index.tsx:75]: Yes, I agree, TS is terrible for not knowing defaultProps. 
Anyway, I think we can with default value when destruction `directions`

File: src/incubator/panView/panningUtil.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/incubator/panView/panningUtil.ts]: Yup, that's the way to do it..

**Comment by ethanshar** [comment]
@M-i-k-e-l 
I think you missed some comments in the PR, I mentioned you in relevant places

File: src/incubator/panView/index.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:51]: Removed

File: src/incubator/panView/index.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:189]: As you wish (although I think it's more user friendly to use the static types).

File: src/incubator/panView/index.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:193]: Not sure (can't see usage ATM, hope it doesn't), removed

File: src/incubator/panView/index.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:75]: Done

File: src/incubator/panView/index.tsx [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+    return {
+      directionLock,
+      translationLock: dismissible ? TranslationLock.NONE : translationLock!,
+      currentTranslation: {x: translationX.value, y: translationY.value}
+    };
+  };
+
+  const setTranslation = (event: PanGestureHandlerEventPayload, initialTranslation: Frame) => {
+    'worklet';
+    const result = getTranslation(event, initialTranslation, directions!, getTranslationOptions());
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/panView/index.tsx:106]: Done

**Comment by M-i-k-e-l** [comment]
Added `springBack` (formerly known as `autoReturn`); I think all comments are covered.

**Review by ethanshar — APPROVED** [review]

**Pull request merged**